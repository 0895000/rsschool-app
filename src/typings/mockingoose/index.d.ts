declare module 'mockingoose' {
  const mockingoose: any;

  export default mockingoose;
}
