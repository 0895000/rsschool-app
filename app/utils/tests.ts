import { IRouterContext } from 'koa-router';
import { ILogger } from '../logger';

export function createTestContext(userId = 'apalchys'): IRouterContext {
    return {
        request: {
            body: undefined,
        },
        state: {
            user: {
                _id: userId,
            },
        },
    } as any;
}

export function getLoggerMock() {
    const logger: ILogger = {
        child: jest.fn(() => logger),
        error: jest.fn(),
        info: jest.fn(),
        warn: jest.fn(),
    };
    return logger;
}
