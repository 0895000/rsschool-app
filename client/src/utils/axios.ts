import { AxiosRequestConfig } from 'axios';
import { NextPageContext } from 'next';
import getConfig from 'next/config';

const { serverRuntimeConfig } = getConfig();

export function getServerAxiosProps(ctx?: NextPageContext): Partial<AxiosRequestConfig> {
  return {
    baseURL: serverRuntimeConfig.rsHost || '',
    headers:
      ctx && ctx.req
        ? {
            cookie: ctx.req.headers.cookie,
          }
        : undefined,
  };
}
