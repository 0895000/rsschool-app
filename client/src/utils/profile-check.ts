export const checkIsProfileOwner = (githubId: string, requestedGithubId: string): boolean => {
  return githubId === requestedGithubId;
};
