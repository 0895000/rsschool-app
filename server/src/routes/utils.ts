import { IApiResponse } from '../models';
import Router from 'koa-router';
import * as crypto from 'crypto';

export function setResponse<T>(ctx: Router.RouterContext, status: number, data?: T) {
  ctx.status = status;
  ctx.body = {
    data,
  } as IApiResponse<T>;
  return ctx;
}

export function setCsvResponse(ctx: Router.RouterContext, status: number, data: string, filename = 'csv') {
  ctx.status = status;
  ctx.body = data;
  ctx.res.setHeader('Content-Type', 'text/csv');
  if (filename) {
    ctx.res.setHeader('Content-disposition', `filename="${filename}.csv"`);
  }
  return ctx;
}

export function createComparisonSignature(body: any, secret: string) {
  const hmac = crypto.createHmac('sha1', secret);
  const selfSignature = hmac.update(JSON.stringify(body)).digest('hex');
  return `sha1=${selfSignature}`;
}

export function compareSignatures(signature: string, comparisonSignature: string) {
  const source = Buffer.from(signature);
  const comparison = Buffer.from(comparisonSignature);
  return crypto.timingSafeEqual(source, comparison);
}
