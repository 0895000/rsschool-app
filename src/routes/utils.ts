import { IApiResponse } from '../models';
import * as Router from 'koa-router';

export function setResponse<T>(ctx: Router.RouterContext, status: number, data?: T) {
  ctx.status = status;
  ctx.body = {
    data,
  } as IApiResponse<T>;
  return ctx;
}

export function setCsvResponse(ctx: Router.RouterContext, status: number, data: string, filename = 'csv') {
  ctx.status = status;
  ctx.body = data;
  ctx.res.setHeader('Content-Type', 'text/csv');
  if (filename) {
    ctx.res.setHeader('Content-disposition', `filename="${filename}.csv"`);
  }
  return ctx;
}
