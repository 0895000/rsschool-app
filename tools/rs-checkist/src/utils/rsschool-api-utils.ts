import nodeFetch from 'node-fetch';
import { Student, Result, Service } from '../interfaces';
import { AUTH_TOKEN, REQUESTS, REQUEST_REPLACES } from '../constants/rsschool-api';
import { filterLogin } from './text-utils';
import logger from './logger';

let courseId: number;

export const setUpCourseId = async (courseAlias: string) => {
  try {
    const courses = await apiRequest(REQUESTS.getCourses);

    const course = courses.data.find((course: any) => course.alias === courseAlias);

    if (course) {
      courseId = course.id;
    } else {
      courseId = courses.data && courses.data[courses.data.length - 1].id || undefined;
    }

    console.log(`\t=> courseId is ${courseId}`);
  } catch (error) {
    console.log('Error with getting courseId was occured:');
    console.log(error);
  }
};

const apiRequest = async (uri: string, param: string = '', body: any = '') => {
  const apiUri = uri.replace(REQUEST_REPLACES.COURSE_ID, String(courseId)) + param;

  const options: any = {
    headers: {
      Authorization: AUTH_TOKEN,
      'Content-Type': 'application/json; charset=utf-8',
    },
  };

  if (body) {
    options.method = 'POST';
    options.body = body;
  }

  const data = await nodeFetch(apiUri, options);

  let result;
  try {
    result = await data.json();
    return result;
  } catch (error) {
    console.log(`Wrong request to ${data.url}`);
    console.log(`${data.status} ${data.statusText}`);
  }
};

const findTaskByName = async (name: string) => {
  const rawTasks = await apiRequest(REQUESTS.getTasks);
  const task = rawTasks.data
    .find((task: any) => task.name.trim().toLowerCase() === name.trim().toLowerCase());

  return task;
};

export const getStudents = async (service: string) => {
  const rawStudents = await apiRequest(REQUESTS.getStudents);

  const students = rawStudents.data.map(({ githubId, externalAccounts = [] }: any) => {
    const foundService = externalAccounts.find((student: any) => student.service === service);
    const rawServiceLogin = foundService ? foundService.username : null;
    const serviceLogin = filterLogin(rawServiceLogin);

    const data: Student = {
      serviceLogin,
      githubLogin: githubId,
    };

    if (serviceLogin !== rawServiceLogin) {
      logger.push({
        description: `Wrong ${service} login: ${rawServiceLogin}`,
        githubLogin: githubId,
        serviceLogin: rawServiceLogin,
      });
    }
    return data;
  });

  return students;
};

export const prepareScores = async (service: Service, results: Result[], altName?: string) => {
  const taskName = altName || service.name;
  const foundTask = await findTaskByName(taskName);
  const courseTaskId = foundTask && foundTask.courseTaskId || null;

  if (!courseTaskId) {
    console.log('\n\nError: Task not found in database!');
  }

  const scores = results.map((student: Result) => ({
    courseTaskId,
    studentGithubId: student.github,
    mentorGithubId: null,
    comment: null,
    githubPrUrl: null,
    score: student.score,
  }));

  return scores;
};

export const updateScores = async (scores: any) => {
  try {
    return await apiRequest(REQUESTS.updateScores, '', scores);
  } catch (error) {
    console.log('Error with update scores was occured:');
    console.log(error);
    return [];
  }
};
