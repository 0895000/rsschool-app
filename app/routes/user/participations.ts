import { NOT_FOUND, OK } from 'http-status-codes';
import * as Router from 'koa-router';
import { userService } from '../../services';
import { setResponse } from '../utils';

export const getParticipationsRoute = async (ctx: Router.IRouterContext) => {
    const userId = ctx.state.user!._id;
    const user = await userService.getUserById(userId);
    if (user === null) {
        ctx.status = NOT_FOUND;
        return;
    }
    setResponse(ctx, OK, user.participations);
};
