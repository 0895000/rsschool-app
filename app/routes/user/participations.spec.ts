import mockingoose from 'mockingoose';
import { createTestContext } from '../../utils';
import { getParticipationsRoute } from './participations';

describe('User Participations route', () => {
    beforeEach(() => {
        mockingoose.resetAll();
    });

    it('returns data', async () => {
        const ctx = createTestContext();
        mockingoose.User.toReturn(
            {
                _id: 'apalchys',
                profile: {},
            },
            'findOne',
        );
        await getParticipationsRoute(ctx);
        expect(ctx.status).toBe(200);
        expect(ctx.body).toMatchSnapshot();
    });
});
