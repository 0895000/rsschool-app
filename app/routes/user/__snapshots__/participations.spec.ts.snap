// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Participations route returns data 1`] = `
Object {
  "data": CoreMongooseArray [],
}
`;
